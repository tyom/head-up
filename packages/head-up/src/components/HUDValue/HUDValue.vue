<template>
  <dl class="HUDValue">
    <dt class="label">{{ label }}</dt>
    <dd class="value">
      <v-icon
        v-if="increase"
        name="caret-up"
        class="change-increase"
      />
      <v-icon
        v-if="decrease"
        name="caret-down"
        class="change-decrease"
      />
      {{ tweenedValue }}
    </dd>
  </dl>
</template>

<script>
import { tween } from 'femtotween';

export default {
  name: 'HUDValue',
  props: {
    label: {
      type: String,
      default: '',
    },
    value: {
      type: [String, Number],
      default: null,
    },
    increase: {
      type: Boolean,
      default: false,
    },
    decrease: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      tweenedValue: this.value || 'N/A',
    };
  },
  watch: {
    value(newVal, oldVal) {
      tween(oldVal, newVal, v => {
        const newNumber = Number(v);
        this.tweenedValue = newNumber ? newNumber.toFixed(3) : v;
      });
    },
  },
};
</script>

<style scoped>
.value {
  display: flex;
  align-items: center;
  justify-content: center;
}

.fa-icon {
  width: 1.2em;
  height: 1.2em;
  margin-right: 0.2em;
}

.change-increase {
  color: #00c200;
}

.change-decrease {
  color: #ff313e;
}
</style>
