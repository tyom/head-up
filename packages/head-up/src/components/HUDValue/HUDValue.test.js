import { shallowMount } from '@vue/test-utils';
import { tween } from 'femtotween';
import HUDValue from './HUDValue';

jest.mock('femtotween', () => ({
  tween: jest.fn(),
}));

test('render default', () => {
  const wrapper = shallowMount(HUDValue);
  expect(wrapper).toMatchSnapshot();
});

test('render with label and value', () => {
  const wrapper = shallowMount(HUDValue, {
    propsData: {
      label: 'Strength',
      value: 100,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render with up trend', () => {
  const wrapper = shallowMount(HUDValue, {
    propsData: {
      label: 'Strength',
      value: 100,
      increase: true,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render with down trend', () => {
  const wrapper = shallowMount(HUDValue, {
    propsData: {
      label: 'Strength',
      value: 100,
      decrease: true,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

describe('tweened value', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(HUDValue, {
      propsData: {
        label: 'Strength',
        value: 100,
      },
    });

    wrapper.setData({
      value: 120,
    });
  });

  test('calls tween function', () => {
    expect(tween).toHaveBeenCalledWith(100, 120, expect.any(Function));
  });

  test('sets tweenedValue data property', () => {
    const [, , cb] = tween.mock.calls[0];

    cb(123.456789);

    expect(wrapper.vm.tweenedValue).toEqual('123.457');

    cb('purple');

    expect(wrapper.vm.tweenedValue).toEqual('purple');
  });
});
