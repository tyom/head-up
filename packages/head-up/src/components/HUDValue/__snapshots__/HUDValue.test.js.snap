// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render default 1`] = `
<dl class="HUDValue"><dt class="label"></dt>
  <dd class="value">
    <!---->
    <!---->
    N/A
  </dd>
</dl>
`;

exports[`render with down trend 1`] = `
<dl class="HUDValue"><dt class="label">Strength</dt>
  <dd class="value">
    <!---->
    <v-icon-stub name="caret-down" class="change-decrease"></v-icon-stub>
    100
  </dd>
</dl>
`;

exports[`render with label and value 1`] = `
<dl class="HUDValue"><dt class="label">Strength</dt>
  <dd class="value">
    <!---->
    <!---->
    100
  </dd>
</dl>
`;

exports[`render with up trend 1`] = `
<dl class="HUDValue"><dt class="label">Strength</dt>
  <dd class="value">
    <v-icon-stub name="caret-up" class="change-increase"></v-icon-stub>
    <!---->
    100
  </dd>
</dl>
`;
